In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
import os
import sys

# TODO: change this to the path to your homework folder
# GOOGLE_DRIVE_PATH_AFTER_MYDRIVE = 'NLP /Final project/ChineseHumorSentiment-master/data'
# GOOGLE_DRIVE_PATH = os.path.join('drive', 'Shared drives', GOOGLE_DRIVE_PATH_AFTER_MYDRIVE)
# print(os.listdir(GOOGLE_DRIVE_PATH))
# sys.path.append(GOOGLE_DRIVE_PATH)

In [3]:
import pandas as pd

In [ ]:
task1_set = pd.read_csv("drive/Shared drives/NLP Final Project/human_label/Human_Label_balanced_test_task1.csv - Copy of balanced_test_task1.csv",encoding='utf8')
task2_set = pd.read_csv("drive/Shared drives/NLP Final Project/human_label/Human_Label_balanced_test_task2.csv - Copy of balanced_test_task2.csv",encoding='utf8')
task1_set = task1_set.drop(columns=['id', 'Unnamed: 0'])
task2_set = task2_set.drop(columns=['id', 'Unnamed: 0'])

In [ ]:
import torch
device = 'cuda' if torch.cuda.is_available() else 'cpu'
print(device)

cpu


# Human Label accuracy

In [ ]:
import numpy as np
gold_label_1 = np.array(task1_set["label"])
gold_label_2 = np.array(task2_set["label"])
human_label_1 = np.array(task1_set["label_human"])
human_label_2 = np.array(task2_set["Label_human"])

In [ ]:
human_label_1[human_label_1 == 9] = 0

In [ ]:
print(type(gold_label_1))

<class 'numpy.ndarray'>


In [ ]:
from sklearn.metrics import f1_score

In [ ]:
human_label_2

array([ 2.,  2.,  2.,  2.,  2.,  2.,  0.,  1.,  2.,  2.,  2.,  1.,  1.,
        1.,  1.,  2.,  1.,  2.,  2.,  2.,  1.,  2.,  1.,  2.,  1.,  0.,
        1.,  0.,  1.,  0.,  0.,  0.,  0.,  1.,  1.,  2.,  0.,  2.,  1.,
        1.,  2.,  2.,  2.,  2.,  0.,  1.,  1.,  1.,  0.,  2.,  1.,  0.,
        0.,  1.,  0.,  2.,  0.,  2.,  2.,  2.,  1.,  2.,  2.,  0.,  0.,
        1.,  2.,  1.,  2.,  2.,  2.,  1.,  1.,  2.,  1.,  0.,  2.,  2.,
        2.,  2.,  2.,  1.,  2.,  2.,  2.,  1.,  0.,  1.,  1.,  0.,  1.,
        1.,  2.,  1.,  1.,  2.,  2.,  1.,  0.,  1.,  2.,  2.,  2.,  2.,
        1.,  2.,  0.,  1.,  1.,  1.,  2.,  2.,  2.,  2.,  2.,  2.,  0.,
        1.,  0.,  0.,  1.,  1.,  2.,  1.,  1.,  2.,  1.,  1.,  2.,  2.,
        2.,  2.,  1.,  2.,  2.,  0.,  0.,  2.,  1.,  1.,  2.,  1.,  2.,
        2.,  1.,  1.,  2.,  2.,  0.,  1.,  1.,  1.,  1.,  1.,  2.,  2.,
        0.,  0.,  2.,  2.,  2.,  2.,  1.,  1.,  1.,  0.,  2.,  1.,  1.,
        2.,  2.,  1.,  2.,  0.,  0.,  2.,  2.,  1.,  1.,  1.,  2

In [ ]:
f1_score(gold_label_2, human_label_2)

ValueError: ignored

In [ ]:
correct_num = 0
for index, value in gold_label_1.items():
  if human_label_1[index] == value:
    correct_num += 1

print(correct_num/len(gold_label_1))

0.7983333333333333


In [ ]:
correct_num_2 = 0
for index, value in gold_label_2.items():
  if human_label_2[index] == value:
    correct_num_2 += 1

print(correct_num_2/len(gold_label_2))

0.3739565943238731


# Humor feature analysis

In [4]:
train_set = pd.read_csv("drive/Shared drives/NLP Final Project/Dataset/balanced_train_task2.csv",encoding='utf8')
# task2_set = pd.read_csv("drive/Shared drives/NLP Final Project/human_label/Human_Label_balanced_test_task2.csv - Copy of balanced_test_task2.csv",encoding='utf8')
train_set = train_set.drop(columns=['id', 'Unnamed: 0'])
# task2_set = task2_set.drop(columns=['id', 'Unnamed: 0'])

In [5]:
zero_num_wife = (train_set["joke"][train_set["label"] == 0]).apply(lambda s: any(x in s for x in ["不", "没","无","非","莫","弗","勿","毋","未","否","别","無","休","难道"])).sum()
one_num_wife = (train_set["joke"][train_set["label"] == 1]).apply(lambda s: any(x in s for x in ["不", "没","无","非","莫","弗","勿","毋","未","否","别","無","休","难道"])).sum()
two_num_wife = (train_set["joke"][train_set["label"] == 2]).apply(lambda s: any(x in s for x in["不", "没","无","非","莫","弗","勿","毋","未","否","别","無","休","难道"])).sum()

print("0:", zero_num_wife, "1:", one_num_wife, "2:", two_num_wife)

0: 2447 1: 2065 2: 2336


In [15]:
reference_name1 = (train_set["joke"][train_set["label"] == 0]).apply(lambda s: any(x in s for x in ['爸','妈','婆','公' ,'学生','老师', '医生','律师','家长','阿凡提'])).sum()
reference_name2 = (train_set["joke"][train_set["label"] == 1]).apply(lambda s: any(x in s for x in['爸','妈','婆','公' ,'学生','老师', '医生','律师','家长','阿凡提'])).sum()
reference_name3 = (train_set["joke"][train_set["label"] == 2]).apply(lambda s: any(x in s for x in['爸','妈','婆','公' ,'学生','老师', '医生','律师','家长','阿凡提'])).sum()

print("0:", reference_name1, "1:", reference_name2, "2:", reference_name3)


0: 1285 1: 1068 2: 1390


1. ： 冒号 0: 2600 1: 1898 2: 2578
2. 、 顿号 0: 213 1: 99 2: 135
3. ['爸','妈','婆','公' ,'学生','老师', '医生','律师','家长','阿凡提']，0: 1285 1: 1068 2: 1390
4. 感叹号，！， 0: 1356 1: 1069 2: 1458
5. 问号，？，0: 1650 1: 1328 2: 1700


In [7]:
train_set_1 = pd.read_csv("drive/Shared drives/NLP Final Project/Dataset/balanced_train_task1.csv",encoding='utf8')

In [ ]:
zero_num_wife = (train_set_1["joke"][train_set_1["label"] == 0]).apply(lambda s: any(x in s for x in ["不", "没","无","非","莫","弗","勿","毋","未","否","别","無","休","难道"])).sum()
one_num_wife = (train_set_1["joke"][train_set_1["label"] == 1]).apply(lambda s: any(x in s for x in ["不", "没","无","非","莫","弗","勿","毋","未","否","别","無","休","难道"])).sum()
# two_num_wife = (train_set_1["joke"][train_set_1["label"] == 2]).apply(lambda s: any(x in s for x in["不", "没","无","非","莫","弗","勿","毋","未","否","别","無","休","难道"])).sum()

print("0:", zero_num_wife, "1:", one_num_wife)

0: 1699 1: 2200


In [10]:
reference_name1 = (train_set_1["joke"][train_set_1["label"] == 0]).apply(lambda s: any(x in s for x in["、"])).sum()
reference_name2 = (train_set_1["joke"][train_set_1["label"] == 1]).apply(lambda s: any(x in s for x in["、"])).sum()
# reference_name3 = (train_set_1["joke"][train_set_1["label"] == 2]).apply(lambda s: any(x in s for x in["、"])).sum()

print("0:", reference_name1, "1:", reference_name2)

0: 0 1: 60


task 1 feature:
1. ： 冒号 0: 290 1: 1809
2. 、 顿号 0: 0 1: 60
3. ['爸','妈','婆','公' ,'学生','老师', '医生','律师','家长','阿凡提']，0: 158 1: 965
4. 感叹号，！， 0: 214 1: 950
5. 问号，？，0: 529 1: 1247


## Preprocessing

In [ ]:
! pip install transformers

     |████████████████████████████████| 3.8 MB 7.1 MB/s 
     |████████████████████████████████| 895 kB 44.3 MB/s 
     |████████████████████████████████| 6.5 MB 41.8 MB/s 
     |████████████████████████████████| 67 kB 5.8 MB/s 
     |████████████████████████████████| 596 kB 32.6 MB/s 
  Attempting uninstall: pyyaml
    Found existing installation: PyYAML 3.13
    Uninstalling PyYAML-3.13:
      Successfully uninstalled PyYAML-3.13


In [ ]:
train_set

joke  label
0     请告诉我，史密斯先生，面试官问道，您还有什么其他您认为值得一提的技能吗？ 的确还有，应聘者谦...      1
1     一辆小汽车急速闯过红灯，刚好被警察拦住。 喂，你难道没有看见红灯吗？ 不，真对不起，我看到红...      1
2     教官子与县丞子厮打，教官子屡负，归而哭诉其母。母 曰：彼家终日吃肉，故恁般强健会打。你家终日...      0
3                                   经验是我们秃顶后大自然赋予我们的梳子。      1
4                         正在进行中 - 它可能很慢，但它很稳定。请耐心等待。谢谢。      1
...                                                 ...    ...
8995  一官员因贪污受贿被判刑，儿子探监时，说自己大学毕业找不到工作。 官员说：会有人来帮你的，我在...      1
8996  某医院急诊室送来一位病人，医生见病人痛得厉害，就给他服了一片止痛药，可是病人仍疼痛不止。 家...      2
8997  小学时候，每天出发前爸爸总会躺在床上提醒我： 校徽，红领巾，书包......天天被他说烦了，...      2
8998  法官问犯人：“你一点法律意识没有怎么不去学点法律知识啊？” 犯人答：“你当我傻啊！知法犯法―...      2
8999  一天，迂公拜访袁洗马（官名，分管图书典籍），看见他手里拿着一本书边看边走，便问是什么书。洗马...      0

[9000 rows x 2 columns]

In [ ]:
from transformers import BertTokenizer

tokenizer = BertTokenizer.from_pretrained('bert-base-chinese')

example_text = train_set.iloc[900, 0]
bert_input = tokenizer(example_text,padding='max_length', max_length = 150, 
                       truncation=True, return_tensors="pt")

Downloading:   0%|          | 0.00/107k [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/29.0 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/624 [00:00<?, ?B/s]

In [ ]:
print(bert_input['input_ids'])
print(bert_input['token_type_ids'])
print(bert_input['attention_mask'])

tensor([[ 101, 7370,  749,  689,  865, 1736,  672, 1912, 8024, 3680,  702,  782,
         6963,  833, 3119, 5815,  800, 3064, 4905, 4638,  691, 6205,  511,  102,
            0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
            0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
            0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
            0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
            0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
            0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
            0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
            0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
            0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
            0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
            0,    0,    0,  

In [ ]:
example_text = tokenizer.decode(bert_input.input_ids[0])

print(example_text)

[CLS] 除 了 业 余 园 丁 外 ， 每 个 人 都 会 收 获 他 播 种 的 东 西 。 [SEP] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD] [PAD]


In [ ]:
import torch
import numpy as np
from transformers import BertTokenizer

tokenizer = BertTokenizer.from_pretrained('bert-base-chinese')
labels = {'true': 1,
          'false': 0
          }

class Dataset(torch.utils.data.Dataset):

    def __init__(self, df):

        self.labels = [label for label in df['label']]
        self.texts = [tokenizer(text, 
                               padding='max_length', max_length = 150, truncation=True,
                                return_tensors="pt") for text in df['joke']]

    def classes(self):
        return self.labels

    def __len__(self):
        return len(self.labels)

    def get_batch_labels(self, idx):
        # Fetch a batch of labels
        return np.array(self.labels[idx])

    def get_batch_texts(self, idx):
        # Fetch a batch of inputs
        return self.texts[idx]

    def __getitem__(self, idx):

        batch_texts = self.get_batch_texts(idx)
        batch_y = self.get_batch_labels(idx)

        return batch_texts, batch_y

In [ ]:
from torch import nn
from transformers import BertModel

class BertClassifier(nn.Module):

    def __init__(self, dropout=0.5):

        super(BertClassifier, self).__init__()

        self.bert = BertModel.from_pretrained('bert-base-chinese')
        # self.dropout = nn.Dropout(dropout)
        self.rnn = nn.LSTM(768, 200, num_layers = 3, device='cuda', batch_first=False, bidirectional=True)
        self.linear = nn.Linear(400, 3)
        self.softmax = nn.Softmax(dim=1)

    def forward(self, input_id, mask):

        _, pooled_output = self.bert(input_ids= input_id, attention_mask=mask,return_dict=False)
        # dropout_output = self.dropout(pooled_output)
        rnn_output, _ = self.rnn(pooled_output.view(len(pooled_output), 1, -1))
        linear_output = self.linear(rnn_output)
        final_layer = (linear_output).reshape((2, 3))
        return final_layer

In [ ]:
from torch.optim import Adam
from tqdm import tqdm

def train(model, train_data, val_data, learning_rate, epochs):

    train, val = Dataset(train_data), Dataset(val_data)

    train_dataloader = torch.utils.data.DataLoader(train, batch_size=2, shuffle=True)
    val_dataloader = torch.utils.data.DataLoader(val, batch_size=2)

    use_cuda = torch.cuda.is_available()
    device = torch.device("cuda" if use_cuda else "cpu")

    criterion = nn.CrossEntropyLoss()
    optimizer = Adam(model.parameters(), lr=learning_rate, weight_decay=1e-3)

    if use_cuda:

            model = model.cuda()
            criterion = criterion.cuda()

    for epoch_num in range(epochs):

            total_acc_train = 0
            total_loss_train = 0

            for train_input, train_label in tqdm(train_dataloader):

                train_label = train_label.to(device)
                mask = train_input['attention_mask'].to(device)
                input_id = train_input['input_ids'].squeeze(1).to(device)

                output = model(input_id, mask)
                
                batch_loss = criterion(output, train_label)
                total_loss_train += batch_loss.item()
                acc = (output.argmax(dim=1) == train_label).sum().item()
                total_acc_train += acc

                model.zero_grad()
                batch_loss.backward()
                optimizer.step()
            
            total_acc_val = 0
            total_loss_val = 0

            with torch.no_grad():

                for val_input, val_label in val_dataloader:

                    val_label = val_label.to(device)
                    mask = val_input['attention_mask'].to(device)
                    input_id = val_input['input_ids'].squeeze(1).to(device)

                    output = model(input_id, mask)

                    batch_loss = criterion(output, val_label)
                    total_loss_val += batch_loss.item()
                    
                    acc = (output.argmax(dim=1) == val_label).sum().item()
                    total_acc_val += acc
            
            print(
                f'Epochs: {epoch_num + 1} | Train Loss: {total_loss_train / len(train_data): .3f} \
                | Train Accuracy: {total_acc_train / len(train_data): .3f} \
                | Val Loss: {total_loss_val / len(val_data): .3f} \
                | Val Accuracy: {total_acc_val / len(val_data): .3f}')
                  
EPOCHS = 10
model = BertClassifier()
LR = 1e-6
              
train(model, train_set, test_set, LR, EPOCHS)

Some weights of the model checkpoint at bert-base-chinese were not used when initializing BertModel: ['cls.predictions.transform.LayerNorm.bias', 'cls.seq_relationship.weight', 'cls.predictions.bias', 'cls.predictions.decoder.weight', 'cls.predictions.transform.dense.bias', 'cls.predictions.transform.LayerNorm.weight', 'cls.predictions.transform.dense.weight', 'cls.seq_relationship.bias']
- This IS expected if you are initializing BertModel from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertModel from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
100%|██████████| 4500/4500 [05:31<00:00, 13.58it/s]


Epochs: 1 | Train Loss:  0.548                 | Train Accuracy:  0.345                 | Val Loss:  0.546                 | Val Accuracy:  0.423


100%|██████████| 4500/4500 [05:10<00:00, 14.51it/s]


Epochs: 2 | Train Loss:  0.543                 | Train Accuracy:  0.471                 | Val Loss:  0.538                 | Val Accuracy:  0.487


100%|██████████| 4500/4500 [05:11<00:00, 14.46it/s]


Epochs: 3 | Train Loss:  0.532                 | Train Accuracy:  0.503                 | Val Loss:  0.528                 | Val Accuracy:  0.471


100%|██████████| 4500/4500 [05:09<00:00, 14.52it/s]


Epochs: 4 | Train Loss:  0.517                 | Train Accuracy:  0.506                 | Val Loss:  0.514                 | Val Accuracy:  0.485


100%|██████████| 4500/4500 [05:09<00:00, 14.55it/s]


Epochs: 5 | Train Loss:  0.502                 | Train Accuracy:  0.516                 | Val Loss:  0.500                 | Val Accuracy:  0.506


100%|██████████| 4500/4500 [05:11<00:00, 14.44it/s]


Epochs: 6 | Train Loss:  0.489                 | Train Accuracy:  0.525                 | Val Loss:  0.492                 | Val Accuracy:  0.508


100%|██████████| 4500/4500 [05:09<00:00, 14.52it/s]


Epochs: 7 | Train Loss:  0.476                 | Train Accuracy:  0.541                 | Val Loss:  0.491                 | Val Accuracy:  0.512


100%|██████████| 4500/4500 [05:09<00:00, 14.56it/s]


Epochs: 8 | Train Loss:  0.465                 | Train Accuracy:  0.566                 | Val Loss:  0.491                 | Val Accuracy:  0.516


100%|██████████| 4500/4500 [05:09<00:00, 14.52it/s]


Epochs: 9 | Train Loss:  0.447                 | Train Accuracy:  0.595                 | Val Loss:  0.493                 | Val Accuracy:  0.524


100%|██████████| 4500/4500 [05:08<00:00, 14.58it/s]


Epochs: 10 | Train Loss:  0.429                 | Train Accuracy:  0.622                 | Val Loss:  0.496                 | Val Accuracy:  0.524


In [ ]:
torch.save(model, "drive/Shared drives/NLP Final Project/Star_model/three_way_bert")